File: apps_hosting/best-btp-finder/src/functions/scrape.js
const cheerio = require('cheerio');
const DataFrame = require('pandas-js').DataFrame;

exports.handler = async function(event, context) {

  const fetchModule = await import('node-fetch');
  const fetch = fetchModule.default;

  try {
    const dataframe = await parseHtmlTableToDataFrame();
    const transformedDataframe = applyDataTransformations(dataframe);
    const bestBtps = findBestBtps(transformedDataframe, 3, 3.5, 1, 10, 99);
    return {
      statusCode: 200,
      body: JSON.stringify(bestBtps),
      headers: {
        'Content-Type': 'application/json'
      }
    };
  } catch (error) {
    console.error("Error: ", error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: error.message }),
      headers: {
        'Content-Type': 'application/json'
      }
    };
  }
}

async function parseHtmlTableToDataFrame() {
  console.log('[scrape.js] parseHtmlTableToDataFrame() function starts')
  const url = 'https://www.rendimentibtp.it/';
  const response = await fetch(url);
  const htmlContent = await response.text();
  const $ = cheerio.load(htmlContent);

  const data = {
    isin: [],
    titolo: [],
    durata: [],
    scadenza: [],
    tasso: [],
    prezzo: [],
    cedolaNetta: []
  };

  $('tr').each((i, row) => {
    if (i === 0) return; // skip header row
    const cells = $(row).find('td');
    data.isin.push($(cells[0]).text().trim());
    data.titolo.push($(cells[1]).text().trim());
    data.durata.push($(cells[2]).text().trim());
    data.scadenza.push($(cells[3]).text().trim());
    data.tasso.push($(cells[4]).text().trim());
    data.prezzo.push($(cells[5]).text().trim());
    data.cedolaNetta.push($(cells[6]).text().trim());
  });

  console.log('[scrape.js] parseHtmlTableToDataFrame() function ends')
  return new DataFrame(data);
}

function applyDataTransformations(df) {
  
  console.log('Input df:')
  console.log(df)

  console.log('[scrape.js] applyDataTransformations() function starts')
  
  df._data['prezzo'] = df._data['prezzo'].map(val => parseFloat(val.replace(',', '.')));
  console.log("[scrape.js] - 'prezzo' mapped correctly")
  
  df._data['durata'] = df._data['durata'].map(val => parseFloat(val.replace(',', '.')));
  console.log("[scrape.js] - 'durata' mapped correctly")
  
  df._data['tasso'] = df._data['tasso'].map(val => parseFloat(val.replace(',', '.').replace('%', '')));
  console.log("[scrape.js] - 'tasso' mapped correctly")
  
  df._data['cedolaNetta'] = df._data['tasso'].map((val, i) => (100 / df._data['prezzo'][i]) * val * 0.875);
  console.log("[scrape.js] - 'cedolaNetta' mapped correctly")
  
  df._data['rendimentoNetto'] = df._data['tasso'].map((val, i) => (((100 / df._data['prezzo'][i]) - 1) * 100 / df._data['durata'][i]) * 0.875 + df._data['cedolaNetta'][i]);
  console.log("[scrape.js] - 'rendimentoNetto' mapped correctly")

  console.log('[scrape.js] applyDataTransformations() function ends')

  console.log('Updated df:')
  console.log(df)

  return df;
}

function findBestBtps(df, minCedola, minRendimento, minDurata, maxDurata, maxPrezzo) {
  
  console.log('[scrape.js] findBestBtps() function starts');

  const numRows = df._data['cedolaNetta'].length; // Assuming all columns have the same number of rows
  
  // Filtering data
  const filteredRows = [];
  for(let i = 0; i < numRows; i++) {
    if(
      df._data['cedolaNetta'][i] > minCedola &&
      df._data['durata'][i] > minDurata &&
      df._data['durata'][i] < maxDurata &&
      df._data['prezzo'][i] < maxPrezzo &&
      df._data['rendimentoNetto'][i] > minRendimento
    ) {
      const row = {};
      Object.keys(df._data).forEach(key => {
        row[key] = df._data[key][i];
      });
      filteredRows.push(row);
    }
  }

  // Sorting by 'rendimentoNetto' in descending order
  const sortedRows = filteredRows.sort((a, b) => b['rendimentoNetto'] - a['rendimentoNetto']);

  console.log('[scrape.js] findBestBtps() function ends');
  
  return sortedRows;
}


// You may want to wrap the `handler` function call in a way that works with your specific environment.

File: apps/best-btp-finder/src/functions/scrape.py
from bs4 import BeautifulSoup
import pandas as pd
import requests
import json

def handler(event, context):
    """
    Handler function for Netlify Function.
    """
    dataframe = parse_html_table_to_dataframe()
    dataframe = apply_data_transformations(dataframe)
    best_btps = found_best_btps(
        dataframe,
        min_cedola=3,
        max_durata=10,
        min_durata=1,
        min_rendimento=3.5,
        max_prezzo=99
        )

    return {
        'statusCode': 200,
        'body': json.dumps(best_btps.to_dict(orient='records')),
        'headers': {
            'Content-Type': 'application/json'
        },
    }


def parse_html_table_to_dataframe():
    """
    Scrapes BTPs information from website and returns as DataFrame.
    """

    # Make a request to the website and get the HTML content
    url = 'https://www.rendimentibtp.it/' # replace with the actual URL of the website you are scraping
    response = requests.get(url, timeout=60)
    html_content = response.content

    # Parse the HTML content with BeautifulSoup
    soup = BeautifulSoup(html_content, 'html.parser')

    # Find all the table rows (tr) in the HTML content
    table_rows = soup.find_all('tr')

    # Create empty lists to store the data for each column
    isins = []
    titoli = []
    durate = []
    scadenze = []
    tassi = []
    prezzi = []
    cedole_netto = []

    # Loop through each table row (skipping the first one, which contains the table headers)
    for row in table_rows[1:]:
        # Find all the table cells (td) in the row
        cells = row.find_all('td')

        # Extract the data from each cell and append it to the corresponding list
        isin = cells[0].a.text
        isins.append(isin)

        titolo = cells[1].text
        titoli.append(titolo)

        durata = cells[2].text
        durate.append(durata)

        scadenza = cells[3].text
        scadenze.append(scadenza)

        tasso = cells[4].text
        tassi.append(tasso)

        prezzo = cells[5].text
        prezzi.append(prezzo)

        cedola_netto = cells[6].text
        cedole_netto.append(cedola_netto)

    data = {
        'isin': isins,
        'titolo': titoli,
        'durata': durate,
        'scadenza': scadenze,
        'tasso': tassi,
        'prezzo': prezzi,
        'cedola netta': cedole_netto
    }
    dataframe = pd.DataFrame(data)
    return dataframe


def apply_data_transformations(dataframe):
    """
    Applies data transformations to the DataFrame
    """
    dataframe['prezzo'] = dataframe['prezzo'].str.replace(',', '.').astype(float)
    dataframe['durata'] = dataframe['durata'].str.replace(',', '.').astype(float)
    dataframe['cedola netta'] = dataframe['cedola netta'].str.replace(',', '.').str.replace('%', '').astype(float)
    dataframe['tasso'] = dataframe['tasso'].str.replace(',', '.').str.replace('%', '').astype(float)
    dataframe['cedola netta'] = 100/dataframe['prezzo'] * dataframe['tasso'] * 0.875
    dataframe['rendimento netto'] = round((((100/dataframe['prezzo'])-1)*100/dataframe['durata'])*0.875 + dataframe['cedola netta'],2)
    dataframe['cedola netta'] = round(dataframe['cedola netta'],2)
    return dataframe


def found_best_btps(dataframe, min_cedola, min_rendimento, min_durata, max_durata, max_prezzo):
    """
    Finds the best BTPs based on criteria and returns filtered DataFrame.
    """
    # filter by cedola netta > min_cedola, durata > min_durata, and durata < max_durata
    filtered_dataframe = dataframe[(dataframe['cedola netta'] > min_cedola) &
                     (dataframe['durata'] > min_durata) &
                     (dataframe['durata'] < max_durata) &
                     (dataframe['prezzo'] < max_prezzo) &
                     (dataframe['rendimento netto'] > min_rendimento)]
    # sort by rendimento netto
    sorted_dataframe = filtered_dataframe.sort_values(by='rendimento netto', ascending=False)

    print(f"💸 Trovati {len(sorted_dataframe)} BTP che soddisfano i tuoi criteri:")
    print(f"    > Rendimento netto >= {min_rendimento}%")
    print(f"    > Cedola netta >= {min_cedola}%")
    print(f"    > Maturity {min_durata}-{max_durata} anni")
    print(f"    > Prezzo <= {max_prezzo}")
    return sorted_dataframe
